### PR TITLE
feat(eslint-plugin): add no-rest-destructuring rule (#6265)

### PR DESCRIPTION
**Pull request by alan910127**
This pull request adds an eslint rule `no-rest-destructuing` to make sure that the query results are not captured by a rest element.

Discussion: #6265 

I referenced the commit for `stable-query-client` rule for what I should do to make a new eslint rule in tanstack/eslint-plugin-query, please inform me if I still missed something. 😄

### CODE REVIEW

**Review by TkDodo — APPROVED** [review]

**Comment by alan910127** [comment]
@TkDodo Thanks for your approval!
Furthermore, as I am not quite familiar with creating eslint rules, I cannot figure out how to write the autofix for this rule. 
Hope someone can help make this rule better by introducing the autofix option.

**Comment by TkDodo** [comment]
I don't think we can or need to auto fix this because we can't really know what it should be auto fixed towards?

**Comment by alan910127** [comment]
That's right, thank you!

File: packages/eslint-plugin-query/src/rules/no-rest-desctructuring/no-rest-destructuring.utils.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/utils'
+import type { TSESTree } from '@typescript-eslint/utils'
+
+export const NoRestDestructuringUtils = {
+  isObjectRestDestructuring(node: TSESTree.Node): boolean {
```
Comment (Newbie012) [inline packages/eslint-plugin-query/src/rules/no-rest-desctructuring/no-rest-destructuring.utils.ts:5]: I would rename it to hasRestElement and move it to ASTUtils, but that's totally fine as well.

**Review by Newbie012 — COMMENTED** [review]
I would add another test case that allows destructing non-query hook with the same name (such as useQuery of chakra-ui).

Other than that, LGTM

File: packages/eslint-plugin-query/src/rules/no-rest-desctructuring/no-rest-destructuring.utils.ts (hunk 1 above)
Comment (alan910127) [reply packages/eslint-plugin-query/src/rules/no-rest-desctructuring/no-rest-destructuring.utils.ts:5]: I think hasRestElement would be confused with the array rest destructuring `[a, ...rest]`, but I'm okay with moving it to ASTUtils.

**Comment by alan910127** [comment]
@Newbie012 I have implemented the test cases for allowing rest destructuring hooks from other packages with the name `useQuery`, `useInfiniteQuery` and `useQueries`.

**Review by Newbie012 — APPROVED** [review]

**Pull request merged**